### PR TITLE
[GORDO-1554] Key generation ASSERT inspection

### PR DESCRIPTION
**Pull request by hkernbach**
3.10 Backport of: https://github.com/arangodb/arangodb/pull/18214
Please see all details there.

**There is also an EE Companion:** 
* https://github.com/arangodb/enterprise/pull/1222

### CODE REVIEW

**Comment by hkernbach** [comment]
Unrelated jenkins failure

**Review by markuspf — APPROVED** [review]
LGTM

File: tests/js/common/shell/shell-document-babies.js [hunk 1]
```diff
@@ -305,30 +306,43 @@ function CollectionDocumentSuiteBabies() {
...
-        assertEqual(docs[1].errorNum, ERRORS.ERROR_ARANGO_DOCUMENT_KEY_BAD.code);
-        collection.remove("a");
+        }];
+
+        if (!cluster) {
+          const docs = collection.insert(batch);
+          assertEqual(docs.length, 2);
+          assertEqual(docs[0]._key, "a");
+          assertTrue(docs[1].error);
+          assertEqual(docs[1].errorNum, ERRORS.ERROR_ARANGO_DOCUMENT_KEY_BAD.code);
```
Comment (dothebart) [inline tests/js/common/shell/shell-document-babies.js:325]: sequence of parameters should be similar; a third parameter with the full document would be even better.

File: tests/js/common/shell/shell-document-babies.js [hunk 2]
```diff
@@ -305,30 +306,43 @@ function CollectionDocumentSuiteBabies() {
...
+          collection.remove("a");
+          assertEqual(collection.count(), 0);
+        } else {
+          try {
+            collection.insert(batch);
```
Comment (jsteemann) [inline tests/js/common/shell/shell-document-babies.js:330]: Wouldn't this different test code mean that we establish a new behavior difference between single server and cluster?
An insert should not have different behavior between single server and cluster. When a batch of documents is inserted, the overall insert should **not** fail even if some/all of the operations in the batch fail. Instead, the caller must check the individual results of each operation.

File: tests/js/common/shell/shell-keygen.js [hunk 3]
```diff
@@ -32,29 +32,246 @@ const jsunity = require("jsunity");
...
+  // single insert, using document API
+  invalidKeys.forEach(invalidKey => {
+    try {
+      db[vn].save({_key: invalidKey, value: "43"});
+    } catch (ignore) {
+    }
+  });
+
+  // batch insert, using document API
+  const batch = [];
```
Comment (jsteemann) [inline tests/js/common/shell/shell-keygen.js:219]: Should `batch` be const here? I know JavaScript is quite special about `const`, but I always keep forgetting what exactly is constant when you declare a JS variable `const`.

File: tests/js/common/shell/shell-keygen.js [hunk 4]
```diff
@@ -32,29 +32,246 @@ const jsunity = require("jsunity");
...
+
+  // Note: Using save() <-> insert() should be the same.
+  // Using both here to double-check that the assumption is
+  // proven.
+
+  // single insert, using document API
+  invalidKeys.forEach(invalidKey => {
+    try {
+      db[vn].insert({_key: invalidKey, value: "43"});
+    } catch (ignore) {
```
Comment (jsteemann) [inline tests/js/common/shell/shell-keygen.js:207]: If this insert attempt is supposed to fail, we should add `fail()` after it for clarity.
I know at the end we are checking the number of docs in the collection, but checking for failures immediately here would be preferrable IMO.

File: tests/js/common/shell/shell-keygen.js (hunk 3 above)
Comment (hkernbach) [reply tests/js/common/shell/shell-keygen.js:219]: Yes it is fine. Explanation can be found here e.g.: https://stackoverflow.com/a/61517230

File: tests/js/common/shell/shell-keygen.js [hunk 5]
```diff
@@ -32,29 +32,246 @@ const jsunity = require("jsunity");
...
+    }
+  });
+  // single insert, using document API
+  invalidKeys.forEach(invalidKey => {
+    try {
+      db[vn].save({_key: invalidKey, value: "43"});
```
Comment (jsteemann) [inline tests/js/common/shell/shell-keygen.js:213]: same here (and for all following occurrences)

File: tests/js/common/shell/shell-keygen.js (hunk 3 above)
Comment (hkernbach) [reply tests/js/common/shell/shell-keygen.js:219]: Basically it means, you cannot change the type of batch. It must stay an array.

**Comment by hkernbach** [comment]
Jenkins green

**Review by jsteemann — APPROVED** [review]
LGTM

**Pull request merged**